### ISSUE TITLE
KPV auto-upgrade via WiFi and possibly 3G

### ISSUE THREAD

**Issue by tigran123** [koreader/kindlepdfviewer#334]
Creating this issue as a proper place for discussion rather than the existing issue #282 which should strictly be about the current release, not some future enhancements....

**Comment by dpavlin** [comment]
Main reason why I suggested packing update inside pdf is to be able to use 3G (only connection on DXG) and built-in browser to get it from Internet. I don't have any idea how hard would it be to extract update from pdf using mupdf.

**Comment by houqp** [comment]
If we pack the file into pdf and want to download it via 3G, we have to manually send the pdf file to our kindle mail, right?

**Comment by tigran123** [comment]
@houqp This is what I wanted @dpavlin to clarify, i.e. I was wondering whether @dpavlin implies that we can wget a pdf file directly via 3G? I know that we can't download it via browser (only txt/prc/mobi files are downloadable, sadly). So, how do we get that PDF file to the device via 3G?

**Comment by dpavlin** [comment]
3G network connectivity requires inserting HTTP headers which are generated when web browser does it's authentication with Amazon's proxy see #221. To be honest, I never managed to get wget working so simplest solution would be to download pdf file using Amazon's web browser.

**Comment by dpavlin** [comment]
@houqp I don't think we need to e-mail document, we can download .mobi and .pdf files directly from web sites (at least on DXG and K3)

**Comment by tigran123** [comment]
@dpavlin .mobi yes, but .pdf not. I.e. not via browser. You can only download txt/prc/mobi files via browser (without FreeDownload hack which I wouldn't recommend anyone to use as it is dangerous and unstable)

**Comment by houqp** [comment]
Does amazon just check the extension or the content of the file?

**Comment by tigran123** [comment]
Their email service definitely checks the content --- I tried sending fb2 and djvu masquerading as pdf and it didn't work, even though it shouldn't look inside my pdf files unless I specified "convert" in the Subject. But the website --- I don't know, we have to check.

**Comment by dpavlin** [comment]
We can then probably pack update inside mobi to deliver it as download from web. It would probably require to embed zip as image and then extract it to disk. I seem to remember methods for that while reading through crengine source a few weeks ago while trying to figure out how to extract links (for which I still don't have solution).

**Comment by tigran123** [comment]
So you are suggesting the user to download it manually and then point KPV to the file? I thought it would be more conventional and convenient if KPV internally went (but ONLY if the user permitted this, not like some nasty software that does things behind the user's back) and checked and downloaded it if necessary. But presumably this (via wget or curl) wouldn't work via 3G because of Amazon handshake secrets etc (which we can't use anyway as it would violate their T&C).

**Comment by dracodoc** [comment]
I think to upgrade through 3G is not realistic, how long will a 3G connection take to download new kpv package? Personally I never use 3G in my DXG except to set the clock.

So I suggest you can just consider wifi only. Although I think the current upgrade method is easy too.
